src/reconcile.py

```diff
@@ -0,0 +1,274 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for calculating required changes based on docs directory and navigation table."""
+
+import itertools
+import typing
+
+from . import exceptions, types_
+from .discourse import Discourse
+
+
+def _local_only(path_info: types_.PathInfo) -> types_.CreateAction:
+    """Return a create action based on information about a local documentation file.
+
+    Args:
+        path_info: Information about the local documentation file.
+
+    Returns:
+        A page create action.
+    """
+    return types_.CreateAction(
+        action=types_.Action.CREATE,
+        level=path_info.level,
+        path=path_info.table_path,
+        navlink_title=path_info.navlink_title,
+        content=path_info.local_path.read_text() if path_info.local_path.is_file() else None,
+    )
+
+
+def _local_and_server(
+    path_info: types_.PathInfo, table_row: types_.TableRow, discourse: Discourse
+) -> (types_.UpdateAction | types_.NoopAction | types_.CreateAction | types_.DeleteAction):
+    """Return an update or noop action depending on if the content or navlink title has changed.
+
+    Args:
+        path_info: Information about the local documentation file.
+        table_row: A row from the navigation table.
+        discourse: A client to the documentation server.
+
+    Returns:
+        A page update or noop action.
+
+    Raises:
+        ReconcilliationError: if the table path or level do not match for the path info and table
+            row.
+    """
+    if path_info.level != table_row.level:
+        raise exceptions.ReconcilliationError(
+            f"internal error, level mismatch, {path_info=!r}, {table_row=!r}"
+        )
+    if path_info.table_path != table_row.path:
+        raise exceptions.ReconcilliationError(
+            f"internal error, table path mismatch, {path_info=!r}, {table_row=!r}"
+        )
+
+    # Is a directory locally and a grouping on the server
+    if path_info.local_path.is_dir() and table_row.is_group:
+        if table_row.navlink.title == path_info.navlink_title:
+            return types_.NoopAction(
+                action=types_.Action.NOOP,
+                level=path_info.level,
+                path=path_info.table_path,
+                navlink=table_row.navlink,
+                content=None,
+            )
+        return types_.UpdateAction(
+            action=types_.Action.UPDATE,
+            level=path_info.level,
+            path=path_info.table_path,
+            navlink_change=types_.NavlinkChange(
+                old=table_row.navlink,
+                new=types_.Navlink(title=path_info.navlink_title, link=table_row.navlink.link),
+            ),
+            content_change=types_.ContentChange(old=None, new=None),
```

Comment:
Yes I think that is clearer, it does make using the action more difficult since now a check on `content_change` is required before using it to ensure it isn't `None` which wasn't required before, it is another one of those edge cases that shouldn't occur in practice

Reply:
This has now been pushed to this PR